### PR TITLE
refactor(android): Re-added "extendSafeArea" support to refactored TabGroup in 8.0.0

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-26677

**Summary:**
This is not a bug.
PR #10358 which refactored `TabGroup` in Titanium 8.0.0 temporarily dropped "extendSafeArea" support. This is a follow-up PR to re-add it and to add a new custom inset provider used to treat the top/bottom tab bar as an inset (ie: exclude tab bar from activity window's returned safe-area) since tabs are no longer implemented via `ActionBar`.

**Note:**
Also fixed a minor bug in `TiCompositeLayout` which did not exclude view's flagged with visibility state `View.GONE` from the layout. This was only a bug with the new 8.0.0 refactored `TabGroup` using a top tab bar style and while `disableTabNavigation(true)` was set.

**Test:**
Test procedure to be written...

```javascript
// Set up tab 1.
var tabWindow1 = Ti.UI.createWindow({
	title: "Tab 1",
	extendSafeArea: true,
	backgroundColor: "red",
});
var safeView1 = Ti.UI.createView();
safeView1.add(Ti.UI.createLabel({
	text: "Tab 1",
	textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
	verticalAlign: Ti.UI.TEXT_VERTICAL_ALIGNMENT_CENTER,
	backgroundColor: "green",
	borderWidth: "4dp",
	borderColor: "blue",
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
}));
tabWindow1.add(safeView1);
tabWindow1.addEventListener("postlayout", function() {
	// Tab window's safe-area padding also adds top/bottom tab bar padding as well.
	safeView1.applyProperties(tabWindow1.safeAreaPadding);
});

// Set up tab 2.
var tabWindow2 = Ti.UI.createWindow({
	title: "Tab 2",
	extendSafeArea: true,
	backgroundColor: "orange",
});
var safeView2 = Ti.UI.createView();
safeView2.add(Ti.UI.createLabel({
	text: "Tab 2",
	textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
	verticalAlign: Ti.UI.TEXT_VERTICAL_ALIGNMENT_CENTER,
	backgroundColor: "green",
	borderWidth: "4dp",
	borderColor: "purple",
	width: Ti.UI.FILL,
	height: Ti.UI.FILL,
}));
tabWindow2.add(safeView2);
tabWindow2.addEventListener("postlayout", function() {
	// Tab window's safe-area padding also adds top/bottom tab bar padding as well.
	safeView2.applyProperties(tabWindow2.safeAreaPadding);
});

// Set up a TabGroup with the above tabs and then display it.
var tabGroupSettings = {
	tabs: [
		Ti.UI.createTab({ window: tabWindow1, title: "Tab 1" }),
		Ti.UI.createTab({ window: tabWindow2, title: "Tab 2" }),
	],
	extendSafeArea: true,
	tabsBackgroundColor: "#88000000",
	tabsBackgroundSelectedColor: "transparent",
};
if (Ti.Platform.name === "android") {
	tabGroupSettings.windowFlags |= Ti.UI.Android.FLAG_TRANSLUCENT_STATUS;
	tabGroupSettings.windowFlags |= Ti.UI.Android.FLAG_TRANSLUCENT_NAVIGATION;
	tabGroupSettings.theme = "Theme.AppCompat.NoTitleBar";
//	tabGroupSettings.theme = "Theme.AppCompat.NoTitleBar.Fullscreen";
//	tabGroupSettings.style = Ti.UI.Android.TABS_STYLE_BOTTOM_NAVIGATION;
}
var tabGroup = Ti.UI.createTabGroup(tabGroupSettings);
tabGroup.addEventListener("open", function(e) {
	// Show/hide tab bar every 2 seconds. This tests safe-area height change.
	if (Ti.Platform.name === "android") {
		var wasDisabled = false;
		setInterval(function() {
			wasDisabled = !wasDisabled;
			tabGroup.disableTabNavigation(wasDisabled);
		}, 2000);
	}
});
tabGroup.open();
```

### CODE REVIEW

File: android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java [hunk 1]
```diff
@@ -1929,4 +1930,32 @@ public Rect getSafeAreaRect()
 		}
 		return this.safeAreaMonitor.getSafeAreaRect();
 	}
+
+	/**
+	 * Adds an object used to provide custom insets to be excldued from the safe-area returned
```
Comment (ypbnv) [inline android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java]: Typo in "excluded"

File: android/titanium/src/java/org/appcelerator/titanium/view/TiActivitySafeAreaMonitor.java [hunk 2]
```diff
@@ -155,6 +172,56 @@ public void setActionBarAddedAsInset(boolean value)
 		update();
 	}
 
+	/**
+	 * Adds an object used to provide custom insets to be excldued from the safe-area returned
```
Comment (ypbnv) [inline android/titanium/src/java/org/appcelerator/titanium/view/TiActivitySafeAreaMonitor.java]: Typo in "excluded".

File: android/titanium/src/java/org/appcelerator/titanium/view/TiInsetsProvider.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2018-Present by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+
+package org.appcelerator.titanium.view;
+
+import android.graphics.Rect;
```
Comment (ypbnv) [inline android/titanium/src/java/org/appcelerator/titanium/view/TiInsetsProvider.java]: Unused import.

**Review by ypbnv — APPROVED** [review]
**CR: Pass**
I have left comments for few minor things.

**Review by ssjsamir — APPROVED** [review]
**FR Passed**  "extendSafeArea" support has been added to the refactored TabGroup in 8.0.0 and the following message is no longer displayed. 

`You cannot use a translucent status bar or navigation bar unless you set the window's 'extendSafeArea' property to true.`

**Test Environment**
Google pixel xl 9 (Simulated notch)
Google pixel xl 7.1.1 sim
APPC CLI: 7.0.9
Operating System Name: Mac OS Mojave
Operating System Version: 10.14.2
Node.js Version: 8.9.1
Xcode 10.1

**Comment by ssjsamir** [comment]
@jquick-axway Could you take a look at the merge conflicts?

**Comment by sgtcoolguy** [comment]
Merged manually.

**Pull request closed without merging**